In [1]:
import myComponents as tc
import topic as tp
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import random
import os

In [2]:
# network parameter
mode = 'SSF'

num_broker = 1000
num_sub = 8000
num_pub = 2000

broker_r = 300
broker_rates = [broker_r]*num_broker
topic_dist = [5,5,5,5,5]
monitor_rate = 1
runtime = 100
connection_style = None

# sub parameter
avg_sub_size = 10.0
sub_r = 3

sub_rates = [sub_r]*num_sub
sub_num_topic = [100]*num_sub
sub_diameter = [3]*num_sub

# pub parameter
avg_pub_size = 20.0
pub_r = 3

pub_rates = [pub_r]*num_pub
pub_num_topic = [100]*num_pub
pub_diameter = [8]*num_pub


wildcard_rate=1
plus_rate=0
hash_rate=0


seed = 1



In [ ]:
dir_name = str(num_broker)+'_'+str(num_sub)+'_'+str(num_pub)+'_'+str(sub_r)+'_'+str(pub_r)+'_'+str(wildcard_rate*plus_rate)+'_'+str(wildcard_rate*hash_rate)
os.makedirs('../new_data/'+dir_name, exist_ok = True)

sub_waits = open('../new_data/'+dir_name+'/'+mode+'_'+'sub_waits.csv', 'w')
sub_pkt = open('../new_data/'+dir_name+'/'+mode+'_'+'sub_pkt.csv', 'w')
broker_queue = open('../new_data/'+dir_name+'/'+mode+'_'+'broker_queue.csv', 'w')
broker_output = open('../new_data/'+dir_name+'/'+mode+'_'+'broker_output.csv', 'w')

tc.SwitchPort.mode = mode
total_topic = tp.TopicTree(wildcard_rate, plus_rate, hash_rate)
total_topic.random_construct(topic_dist, seed)
#total_topic.visualize(total_topic.root)
net = tc.Network(total_topic, avg_sub_size, avg_pub_size, sub_waits, sub_pkt, broker_queue, broker_output, qlimit=None)
net.initialize_nodes(broker_rates, sub_rates, sub_num_topic, sub_diameter, pub_rates, pub_num_topic, pub_diameter, monitor_rate, seed)
net.establish_topology(seed)
net.connect_client(connection_style, seed)
print(tc.SwitchPort.mode)
    
net.env.run(runtime)
    
sub_waits.close()
sub_pkt.close()
broker_queue.close()
broker_output.close()

initializing brokers
broker_id:  1
broker_id:  11
broker_id:  21
broker_id:  31
broker_id:  41
broker_id:  51
broker_id:  61
broker_id:  71
broker_id:  81
broker_id:  91
broker_id:  101
broker_id:  111
broker_id:  121
broker_id:  131
broker_id:  141
broker_id:  151
broker_id:  161
broker_id:  171
broker_id:  181
broker_id:  191
broker_id:  201
broker_id:  211
broker_id:  221
broker_id:  231
broker_id:  241
broker_id:  251
broker_id:  261
broker_id:  271
broker_id:  281
broker_id:  291
broker_id:  301
broker_id:  311
broker_id:  321
broker_id:  331
broker_id:  341
broker_id:  351
broker_id:  361
broker_id:  371
broker_id:  381
broker_id:  391
broker_id:  401
broker_id:  411
broker_id:  421
broker_id:  431
broker_id:  441
broker_id:  451
broker_id:  461
broker_id:  471
broker_id:  481
broker_id:  491
broker_id:  501
broker_id:  511
broker_id:  521
broker_id:  531
broker_id:  541
broker_id:  551
broker_id:  561
broker_id:  571
broker_id:  581
broker_id:  591
broker_id:  601
broker_id:  61

sub_id:  4771
sub_id:  4781
sub_id:  4791


In [11]:
wildcard_rate=1
plus_rate=0.1
hash_rate=0.01
dir_name = str(num_broker)+'_'+str(num_sub)+'_'+str(num_pub)+'_'+str(sub_r)+'_'+str(pub_r)+'_'+str(wildcard_rate*plus_rate)+'_'+str(wildcard_rate*hash_rate)
os.makedirs('../data/'+dir_name, exist_ok = True)

sub_waits = open('../data/'+dir_name+'/'+mode+'_'+'sub_waits.csv', 'w')
sub_pkt = open('../data/'+dir_name+'/'+mode+'_'+'sub_pkt.csv', 'w')
broker_queue = open('../data/'+dir_name+'/'+mode+'_'+'broker_queue.csv', 'w')
broker_output = open('../data/'+dir_name+'/'+mode+'_'+'broker_output.csv', 'w')

tc.SwitchPort.mode = mode
total_topic = tp.TopicTree(wildcard_rate, plus_rate, hash_rate)
total_topic.random_construct(topic_dist, seed)
#total_topic.visualize(total_topic.root)
net = tc.Network(total_topic, avg_sub_size, avg_pub_size, sub_waits, sub_pkt, broker_queue, broker_output, qlimit=None)
net.initialize_nodes(broker_rates, sub_rates, sub_num_topic, sub_diameter, pub_rates, pub_num_topic, pub_diameter, monitor_rate, seed)
net.establish_topology(seed)
net.connect_client(connection_style, seed)
print(tc.SwitchPort.mode)
    
net.env.run(runtime)
    
sub_waits.close()
sub_pkt.close()
broker_queue.close()
broker_output.close()

SSF
current time:  5
current time:  10
current time:  15
current time:  20
current time:  25
current time:  30
current time:  35
current time:  40
current time:  45
current time:  50
current time:  55
current time:  60
current time:  65
current time:  70
current time:  75
current time:  80
current time:  85
current time:  90
current time:  95


In [3]:
wildcard_rate=1
plus_rate=0.3
hash_rate=0.03
dir_name = str(num_broker)+'_'+str(num_sub)+'_'+str(num_pub)+'_'+str(sub_r)+'_'+str(pub_r)+'_'+str(wildcard_rate*plus_rate)+'_'+str(wildcard_rate*hash_rate)
os.makedirs('../data/'+dir_name, exist_ok = True)

sub_waits = open('../data/'+dir_name+'/'+mode+'_'+'sub_waits.csv', 'w')
sub_pkt = open('../data/'+dir_name+'/'+mode+'_'+'sub_pkt.csv', 'w')
broker_queue = open('../data/'+dir_name+'/'+mode+'_'+'broker_queue.csv', 'w')
broker_output = open('../data/'+dir_name+'/'+mode+'_'+'broker_output.csv', 'w')

tc.SwitchPort.mode = mode
total_topic = tp.TopicTree(wildcard_rate, plus_rate, hash_rate)
total_topic.random_construct(topic_dist, seed)
#total_topic.visualize(total_topic.root)
net = tc.Network(total_topic, avg_sub_size, avg_pub_size, sub_waits, sub_pkt, broker_queue, broker_output, qlimit=None)
net.initialize_nodes(broker_rates, sub_rates, sub_num_topic, sub_diameter, pub_rates, pub_num_topic, pub_diameter, monitor_rate, seed)
net.establish_topology(seed)
net.connect_client(connection_style, seed)
print(tc.SwitchPort.mode)
    
net.env.run(runtime)
    
sub_waits.close()
sub_pkt.close()
broker_queue.close()
broker_output.close()

SSF
current time:  5
current time:  10
current time:  15
current time:  20
current time:  25
current time:  30
current time:  35
current time:  40
current time:  45
current time:  50
current time:  55
current time:  60
current time:  65
current time:  70
current time:  75
current time:  80
current time:  85
current time:  90
current time:  95


In [4]:
wildcard_rate=1
plus_rate=0.5
hash_rate=0.05
dir_name = str(num_broker)+'_'+str(num_sub)+'_'+str(num_pub)+'_'+str(sub_r)+'_'+str(pub_r)+'_'+str(wildcard_rate*plus_rate)+'_'+str(wildcard_rate*hash_rate)
os.makedirs('../data/'+dir_name, exist_ok = True)

sub_waits = open('../data/'+dir_name+'/'+mode+'_'+'sub_waits.csv', 'w')
sub_pkt = open('../data/'+dir_name+'/'+mode+'_'+'sub_pkt.csv', 'w')
broker_queue = open('../data/'+dir_name+'/'+mode+'_'+'broker_queue.csv', 'w')
broker_output = open('../data/'+dir_name+'/'+mode+'_'+'broker_output.csv', 'w')

tc.SwitchPort.mode = mode
total_topic = tp.TopicTree(wildcard_rate, plus_rate, hash_rate)
total_topic.random_construct(topic_dist, seed)
#total_topic.visualize(total_topic.root)
net = tc.Network(total_topic, avg_sub_size, avg_pub_size, sub_waits, sub_pkt, broker_queue, broker_output, qlimit=None)
net.initialize_nodes(broker_rates, sub_rates, sub_num_topic, sub_diameter, pub_rates, pub_num_topic, pub_diameter, monitor_rate, seed)
net.establish_topology(seed)
net.connect_client(connection_style, seed)
print(tc.SwitchPort.mode)
    
net.env.run(runtime)
    
sub_waits.close()
sub_pkt.close()
broker_queue.close()
broker_output.close()

KeyboardInterrupt: 

In [6]:
a = [1,12,4]
k=random.choice(a)
a.remove(k)
print(k)
print(a)

12
[1, 4]


# plot
net.visualize()
plt.figure(figsize=(10, 3), dpi=80)
#monitor.queue_size
#monitor.queue_length
for i in range(len(net.broker_monitor_list)):
    monitor = net.broker_monitor_list[i]
    plt.plot(monitor.queue_length, label=monitor.port.sp_id)
plt.legend(bbox_to_anchor=(1.01, 1),loc='upper left')
plt.title("queue length")
window = 6
plt.figure(figsize=(10, 3), dpi=80)
for i in range(len(net.broker_monitor_list)):
    monitor = net.broker_monitor_list[i]
    a = np.array(monitor.bytes_sent)
    b = np.array([0]*window+monitor.bytes_sent[0:-window])
    plt.plot((a-b)[window-1:]/window/monitor_rate, label=monitor.port.sp_id)
plt.legend(bbox_to_anchor=(1.01, 1),loc='upper left')
plt.title("output rate in byte/s, window="+str(window))
import warnings
warnings.filterwarnings('ignore')
window = 6
plt.figure(figsize=(10, 9), dpi=80)

for i in range(len(net.sub_monitor_list)):
    monitor = net.sub_monitor_list[i]
    a = np.array(monitor.packets_rec)
    b = np.array([0]*window+monitor.packets_rec[0:-window])
    avg_packet_rec_per_time = (a-b)[window-1:]/window/monitor_rate
    
    a = np.array(monitor.tot_waits)
    b = np.array([0]*window+monitor.tot_waits[0:-window])
    avg_waits_per_time = (a-b)[window-1:]/window/monitor_rate
    

    plt.subplot(311)
    plt.plot(avg_waits_per_time, label=monitor.client.client_id)
    plt.subplot(312)
    plt.plot(avg_packet_rec_per_time, label=monitor.client.client_id)
    plt.subplot(313)
    plt.plot(avg_waits_per_time/avg_packet_rec_per_time, label=monitor.client.client_id)

plt.subplot(311)
plt.title("avg_waits s/s, window="+str(window))
plt.subplot(312)
plt.title("avg_pkt_recieved /s, window="+str(window))
plt.subplot(313)
plt.title("avg_waits/pkt, window="+str(window))
plt.legend(bbox_to_anchor=(1.01, 1),loc='upper left')
plt.subplots_adjust(top=0.92, bottom=0.08, left=0.10, right=0.95, hspace=0.25,wspace=0.35)